src_c/camera_windows.c

```diff
@@ -0,0 +1,859 @@
+    if (!devices) {
+        hr = E_OUTOFMEMORY;
+        HANDLEHR(hr);
+    }
+
+    for (int i=0; i<(int)count; i++) {
+        if(!_get_name_from_activate(ppDevices[i], &devices[i])) {
+            hr = HR_UPSTREAM_FAILURE;
+            goto cleanup;
+        }
+    }
+
+    *num_devices = count;
+    goto cleanup;
+
+    cleanup:
+        RELEASE(pAttributes);
+        for (int i=0; i<(int)count; i++) {
+            RELEASE(ppDevices[i]);
+        }
+        CoTaskMemFree(ppDevices);
+        if FAILED(hr) {
+            for (int i=0; i<(int)count; i++) {
+                free(devices[i]);
+            }
+            free(devices);
+            return NULL;
+        }
+        else {
+            return devices;
+        }
+}
+
+IMFActivate *
+windows_device_from_name(WCHAR* device_name)
+{
+    IMFAttributes *pAttributes = NULL;
+    IMFActivate **ppDevices = NULL;
+    WCHAR* _device_name = NULL;
+    UINT32 count = 0;
+    HRESULT hr;
+
+    hr = MFCreateAttributes(&pAttributes, 1);
+    HANDLEHR(hr);
+
+    hr = pAttributes->lpVtbl->SetGUID(pAttributes,
+                                      &MF_DEVSOURCE_ATTRIBUTE_SOURCE_TYPE,
+                                      &DEVSOURCE_VIDCAP_GUID);
+    HANDLEHR(hr);
+
+    hr = MFEnumDeviceSources(pAttributes, &ppDevices, &count);
+    HANDLEHR(hr);
+
+    for (int i=0; i<(int)count; i++) {
+        if (!_get_name_from_activate(ppDevices[i], &_device_name)) {
+            hr = HR_UPSTREAM_FAILURE;
+            goto cleanup;
+        }
+        if (!wcscmp(_device_name, device_name)) {
+            free(_device_name);
+            for (int j=0; j<(int)count; j++) {
+                if (i != j) {
+                    RELEASE(ppDevices[j]);
+                }
+            }
+            CoTaskMemFree(ppDevices);
+            return ppDevices[i];
+        }
+        free(_device_name);
+    }
+
+    goto cleanup;
+
+    cleanup:
+        RELEASE(pAttributes);
+        for (int i=0; i<(int)count; i++) {
+            RELEASE(ppDevices[i]);
+        }
+        CoTaskMemFree(ppDevices);
+        return NULL;
+}
+
+int
+_select_source_type(pgCameraObject* self, IMFMediaType **mp)
+{
+    HRESULT hr;
+    IMFMediaType* media_type = NULL;
+    int type_count = 0;
+    UINT32 t_width, t_height;
+    IMFMediaType** native_types = NULL;
+    int *diagonal_distances = NULL;
+    UINT64 size;
+    GUID subtype;
+    int difference = 100000000;
+    int current_difference;
+    int index = -1;
+    int requested_diagonal = DISTANCE(self->width, self->height);
+    UINT64 fps_max_ratio;
+
+    /* Find out how many native media types there are
+     * (different resolution modes, mostly) */
+    while(1) {
+        hr = self->reader->lpVtbl->GetNativeMediaType(self->reader,
+                                                      FIRST_VIDEO,
+                                                      type_count,
+                                                      &media_type);
+        if (hr) {
```

Comment:
Maybe this should reference S_OK to be a bit clearer?
https://docs.microsoft.com/en-us/windows/win32/api/mfreadwrite/nf-mfreadwrite-imfsourcereader-getnativemediatype#return-value

Reply:
I just fed it into the `FAILED` macro, which also resolved one of the warnings.